arwen/elrondapi/manBufOps.go

```diff
@@ -141,75 +164,163 @@ func v1_4_mBufferNewFromBytes(context unsafe.Pointer, dataOffset int32, dataLeng
+//export v1_4_mBufferCopyByteSlice
+func v1_4_mBufferCopyByteSlice(context unsafe.Pointer, sourceHandle int32, startingPosition int32, sliceLength int32, destinationHandle int32) int32 {
 	managedType := arwen.GetManagedTypesContext(context)
 	runtime := arwen.GetRuntimeContext(context)
 	metering := arwen.GetMeteringContext(context)
 
 	gasToUse := metering.GasSchedule().ManagedBufferAPICost.MBufferGetBytes
 	metering.UseGas(gasToUse)
 
-	managedBuffer, err := managedType.GetBytes(mBufferHandle)
+	sourceBytes, err := managedType.GetBytes(sourceHandle)
 	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
 		return 1
 	}
-	managedType.ConsumeGasForThisIntNumberOfBytes(len(managedBuffer))
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(sourceBytes))
+
+	if startingPosition < 0 || sliceLength < 0 || int(startingPosition+sliceLength) > len(sourceBytes) {
+		// does not fail execution if slice exceeds bounds
+		return 1
+	}
+
+	slice := sourceBytes[startingPosition : startingPosition+sliceLength]
+	managedType.SetBytes(destinationHandle, slice)
+
+	gasToUse = math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(len(slice)))
+	metering.UseGas(gasToUse)
+
+	return 0
+}
+
+//export v1_4_mBufferSetBytes
+func v1_4_mBufferSetBytes(context unsafe.Pointer, mBufferHandle int32, dataOffset int32, dataLength int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ManagedBufferAPICost.MBufferSetBytes
+	metering.UseGas(gasToUse)
+	managedType.ConsumeGasForThisIntNumberOfBytes(int(dataLength))
 
-	err = runtime.MemStore(resultOffset, managedBuffer)
+	data, err := runtime.MemLoad(dataOffset, dataLength)
 	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
 		return 1
 	}
 
-	gasToUse = math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(len(managedBuffer)))
+	managedType.SetBytes(mBufferHandle, data)
+
+	gasToUse = math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(len(data)))
 	metering.UseGas(gasToUse)
 
 	return 0
 }
 
 //export v1_4_mBufferAppend
-func v1_4_mBufferAppend(context unsafe.Pointer, mBufferHandle int32, dataOffset int32, dataLength int32) int32 {
+func v1_4_mBufferAppend(context unsafe.Pointer, accumulatorHandle int32, dataHandle int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ManagedBufferAPICost.MBufferSetBytes
+	metering.UseGas(gasToUse)
+
+	dataBufferBytes, err := managedType.GetBytes(dataHandle)
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
+		return 1
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(dataBufferBytes))
```

Comment:
append should cost more than a simple setbytes. as there is at least one operation in plus

Reply:
we have below:
```
       gasToUse = math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(len(data)))
	metering.UseGas(gasToUse)
```
Do we need more than that?

Reply:
i think the base gasCost should be MBufferSetBytes + MBufferGetBytes, than use the extra DataCopyPerByte